File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import defaultdict, deque

def solution(n, computers):
    answer = 0
    graph = defaultdict(list)
    for i, computer in enumerate(computers):
        for j, is_connect in enumerate(computer):
            if is_connect and i < j:
                graph[i].append(j)
                graph[j].append(i)
    nodes = [i for i in range(n)]
    
    def bfs(node):
        dq = deque()
        dq.append(node)
        visited = [0] * n
        visited[node] = 1
        nodes.remove(node)

        while dq:
            c_node = dq.popleft()
            for n_node in graph[c_node]:
                if visited[n_node]:
                    continue
                dq.append(n_node)
                visited[n_node] = 1
                nodes.remove(n_node)
    
    while nodes:
        answer += 1
        bfs(nodes[0])
            
    return answer

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;
class Solution {
    public int solution(int n, int[][] computers) {
        int answer = 0;
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int i = 0; i < n; i++) {
            graph.put(i, new ArrayList<>());
        }
        
        for (int i = 0; i < computers.length; i++) {
            for (int j = i + 1; j < computers[i].length; j++) {
                if (computers[i][j] == 1) {
                    graph.get(i).add(j);
                    graph.get(j).add(i);
                }
            }
        }
        
        List<Integer> nodes = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            nodes.add(i);
        }
        
        while (!nodes.isEmpty()) {
            answer++;
            bfs(nodes.get(0), graph, nodes, n);
        }
        
        return answer;
    }
    
    private void bfs(int node, Map<Integer, List<Integer>> graph, List<Integer> nodes, int n) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(node);
        boolean[] visited = new boolean[n];
        visited[node] = true;
        nodes.remove(Integer.valueOf(node));
        
        while (!queue.isEmpty()) {
            int currentNode = queue.poll();
            for (int nextNode : graph.get(currentNode)) {
                if (visited[nextNode]) continue;
                
                queue.add(nextNode);
                visited[nextNode] = true;
                nodes.remove(Integer.valueOf(nextNode));
            }
        }
    }
}